src/Microsoft.ServiceFabric.Actors/Runtime/KvsActorStateProviderBase.cs

```diff
@@ -1375,15 +1375,109 @@ private Task<IEnumerable<string>> GetStateNamesAsync(ActorId actorId, Cancellati
+            {
+                long previousActorCount = 0L;
+                if (long.TryParse((string)continuationToken.Marker, out previousActorCount))
+                {
+                    enumerator = replica.Enumerate(txn, ActorStateProviderHelper.ActorPresenceStorageKeyPrefix, true);
+                    enumHasMoreEntries = this.actorStateProviderHelper.GetContinuationPointByActorCount(previousActorCount, enumerator, cancellationToken);
+                    enumHasMoreEntries = enumerator.MoveNext();
+                }
+                else
+                {
+                    string lastSeenActorStorageKey = continuationToken.Marker.ToString();
+                    enumerator = replica.Enumerate(txn, lastSeenActorStorageKey, false);
+                    enumHasMoreEntries = enumerator.MoveNext();
+                    var storageKey = enumerator.Current.Metadata.Key;
+                    if (enumHasMoreEntries && storageKey == lastSeenActorStorageKey)
+                    {
+                        enumHasMoreEntries = enumerator.MoveNext();
+                    }
+                }
+
+                if (!enumHasMoreEntries)
+                {
+                    // We are here means the current snapshot that enumerator represents
+                    // has less entries that what ContinuationToken contains.
+                    return Task.FromResult(actorQueryResult);
+                }
+            }
+            else
+            {
+                enumerator = replica.Enumerate(txn, ActorStateProviderHelper.ActorPresenceStorageKeyPrefix, true);
+                enumHasMoreEntries = enumerator.MoveNext();
+            }
+
+            if (!enumHasMoreEntries)
+            {
+                return Task.FromResult(actorQueryResult);
+            }
+
+            while (enumHasMoreEntries && enumerator.Current.Metadata.Key.StartsWith(ActorStateProviderHelper.ActorPresenceStorageKeyPrefix))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var storageKey = enumerator.Current.Metadata.Key;
+                var actorId = ActorStateProviderHelper.GetActorIdFromPresenceStorageKey(storageKey);
+
+                if (actorId != null)
+                {
+                    actorIdList.Add(actorId);
+                }
+                else
+                {
+                    ActorTrace.Source.WriteWarningWithId(
+                        TraceType,
+                        this.traceId,
+                        string.Format("Failed to parse ActorId from storage key: {0}", storageKey));
+                }
+
+                enumHasMoreEntries = enumerator.MoveNext();
+
+                if (actorIdList.Count == itemsCount)
+                {
+                    actorQueryResult.Items = actorIdList.AsReadOnly();
+
+                    // If enumerator has more elements, then set the continuation token.
+                    if (enumHasMoreEntries)

```

Comment:
the condition should be "enumHasMoreEntries && enumerator.Current.Metadata.Key.StartsWith(ActorStateProviderHelper.ActorPresenceStorageKeyPrefix)" to avoid unnecessary in the case  where the last presence key and page count collides.

Reply:
Updated